### PR TITLE
fix: messages content overlap when bottom sheet is shown

### PR DESCRIPTION
**Pull request by kirillzyusko**
### Details

A successor of https://github.com/Expensify/App/pull/16356

The first step of [react-native-keyboard-controller migration plan](https://github.com/Expensify/App/issues/37203)

> [!Important]
> This PR relies on `react-native-keyboard-controller` (`useKeyboardHandler`) hook and `KeyboardAvoidingView` component).

#### How it works 🤔

This PR is **iOS only** for now. Other platforms are not affected. Also, in this explanation, we are talking about iOS only since on Android, iOS web and Android Web keyboard handling works differently.

> [!Note]
> We're talking about default platform behavior. If we start to use `react-native-keyboard-controller` on Android - it'll disable all default keyboard handling by OS and will delegate the keyboard handling to us - so we will have identical behavior across both iOS/Android platforms.
>
> But it may introduce additional issues (because we need to enter edge-to-edge mode - that would be good to do on entire app level, but it may bring some issues with `StatusBar` paddings, so everything has to be verified very carefully). That's why I think it would be better to handle in separate PR.

When we long press on a message, we want to be able to see it. It's easy when we have the keyboard closed. But gets tricky when the keyboard is open.

But first of all let's focus on the case when the keyboard is closed. User long press on the message, we render the bottom sheet (action sheet) with some options. If the message is higher than the height of bottom sheet - the message is visible. When the message is the last one (first from the bottom), it can be fully or partially covered by the bottom sheet.

<details>
<summary>Press to see videos</summary>

|Message above bottom sheet | Message and bottom sheet overlap|
|------------------------------|------------------------------------|
|<video src="https://github.com/Expensify/App/assets/22820318/d36461b1-e03d-4577-ad2a-c57a3ada2749">|<video src="https://github.com/Expensify/App/assets/22820318/4b894ab4-dbe4-4daa-ade9-f44ec42ccf3e">|

</details>

Also everything is tricky when we have keyboard open 😓

<details>
<summary>Press to see videos with keyboard</summary>

|emoji picker instead of the keyboard (the messages should not jump) | showing a message which is covered by the keyboard and the bottom sheet |
|---------------------------------------------------------------------|---------------------------------------------------------------------------------|
|<video src="https://github.com/Expensify/App/assets/22820318/adccd5a8-77ae-4ff6-af2d-ac07fb1f05bf">|<video src="https://github.com/Expensify/App/assets/22820318/765ffead-cbde-47cb-9778-792a196d851e">|

</details>

To achieve this goal we had to:

- use our own scroll component on iOS for `FlatList` of `ReportListView` (it uses reanimated to move the view by the offset)
- use `KeyboardAvoiding` so we're in sync with keyboard-avoiding style updates
- implement animated style handling using a state machine
- create `ActionSheetAwareScrollViewContext` to pass the transition function, so we can transition from one state to another by action triggered from different parts of the application

Below you can find more use cases that involves keyboard and other UI elements interactions:

<details>
<summary>Interactions with other UI elements</summary>

|Add attachment | Call popover|
|----------------|--------------|
|<video src="https://github.com/Expensify/App/assets/22820318/167156b6-29da-4af4-a5a1-9a077248ad42">|<video src="https://github.com/Expensify/App/assets/22820318/572a7053-65df-4bbc-8e0c-df36d9414b2a">|

</details>

> [!Note]
> **Call popover**, **bottom sheet** and **emoji picker** are implemented as **Modal** windows and on iOS keyboard instantly gets hidden when modal window is shown. But "Attachment" popup is implemented not as modal, so for this type of transition we are using `progress`-based animations. When we press on plus -> popup is already shown under keyboard and we dismiss the keyboard. Overall the mechanism is similar to what we've used in the past, except the moment where we substitute `translationY` value - in case when keyboard disappear/appears we interpolate it by `progress` variable to achieve smooth transitions.

### Technical details ⚒️

So in order to calculate the offset for the message, we:

1. Measure the message item in the window using `ref.current.measureInWindow`, so we can have `height` and `y` coordinate of the image
2. Measure the height of the bottom sheet using `onLayout` callback
3. Get the window height from `Dimensions`
4. Get `safeArea.top` since `y` coordinate doesn't get into account safe area
5. Calculate offset as `elementOffset = (y + safeArea.top + itemHeight) - (windowHeight - popoverHeight)`
In order to animate item when they keyboard was open, we need to also know its height which is provided by useAnimatedKeyboard hook from react-native-reanimated, but we should always subtract safeArea.bottom from it since it's not part of the calculation for the offset, but only keyboardHeight.

Overall, the keyboard is the tricky part. On iOS, the keyboard doesn't resize the viewport. So for the content to not be covered by the keyboard, `KeyboardAvoidingView` is used. In React Native, `KeyboardAvoidingView` is implemented as a view that changes its height or resizes the children's view by the keyboard height. It's done using RN's Layout Animations. This results in the animation for the keyboard being applied using the native event emitter callback which uses the bridge. So open/close events of the keyboard will always be delayed that resulting in the delayed keyboard avoiding animation and, which is more important, scheduling layout animations.

In order to be consistent 100% of the time with animation and timings, we had to:

- use `KeyboardAvoidingView` from `react-native-keyboard-controller` for iOS. It reacts on keyboard height/state change on UI thread in the same frame, so we can schedule an update for styles in the next frame
- use just `useDerivedValue` for all the logic, since introducing `useAnimatedReaction` or having multiple shared values will delay updates for 1 frame, but we want to be 100% in sync with keyboard updates
- use `paddingTop: translateY` (because `ScrollView` is inverted).

Another important aspect of the implementation is the usage of the state machine for the animation. State machine allows us to specify the chain of events and how to handle them, and more importantly what actions to ignore, for example:

1. Initial state is idle. it can react to `KEYBOARD_OPEN` action
2. we open emoji picker. it sends` EMOJI_PICKER_OPEN` action
3. there is no handling for this action in idle state so we do nothing
4. we close emoji picker and it sends `EMOJI_PICKER_CLOSE` action which again does nothing
5. we open keyboard. it sends `KEYBOARD_OPEN` action. idle can react to this action by transitioning into keyboardOpen state
6. our state is keyboardOpen. it can react to `KEYBOARD_CLOSE`, `EMOJI_PICKER_OPEN` actions
7. we open emoji picker again. it sends `EMOJI_PICKER_OPEN` action which transitions our state into `emojiPickerOpen` state. now we react only to `EMOJI_PICKER_CLOSE` action
8. before rendering the emoji picker, the app hides the keyboard. it sends `KEYBOARD_CLOSE` action. but we ignore it since our `emojiPickerOpen` state can only handle `EMOJI_PICKER_CLOSE` action. so we write the logic for handling 9. hiding the keyboard but maintaining the offset based on the keyboard state shared value
9. we close the picker and send `EMOJI_PICKER_CLOSE` action which transitions us back into `keyboardOpen` state.

### Fixed Issues
$ https://github.com/Expensify/App/issues/10632
PROPOSAL: N/A

### Tests

These steps were performed on all the platforms. The new behavior is iOS only for now.

1. Open chat
2. Long press on the first message - message should be pushed up to be displayed above the bottom sheet
3. Long press on the top message - if the message is not covered by the bottom sheet, it should stay in place
4. Long press on any attachment (document, message) should result in the same behavior
5. Pressing on reaction icon on the message should push the message above it too.
6. Long press on the message and then pressing on the reactions icon should open the reactions sheet and if the message is covered by the emoji picker it should animate the message so it's above the picker too

Before each next we make sure the keyboard is open:

1. Long press on the message should either animate the message above the bottom sheet or keep it in place
2. Messages should not "jump" because of keyboard hide/open when showing bottom sheet, emoji picker
3. Press on emoji picker in the composer, messages should stay in place even though the keyboard was closer - emoji picker is rendered instead of keyboard
4. Long press on the mesage, click on add reactions - close emoji reactions

- [ ] Verify that no errors appear in the JS console

### Offline tests

Should not affect the offline state.

### QA Steps

Same as Tests

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/22820318/8d85d789-96fa-46fe-8e3c-3b1134557510

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/22820318/5ddc8d16-b47a-4be4-a268-d091296eff82

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/22820318/d53576ee-3e8f-4658-9926-06e5451b8ab8

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/22820318/0670358f-afb4-4770-8862-87bc1fbfc96f

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/22820318/9c5f5877-7a2f-4e75-8ff8-705f612e172c

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/22820318/5abc36ca-b030-45ba-b315-fa1b8ba8f151

</details>

### CODE REVIEW

**Comment by Santhosh-Sellavel** [comment]
@kirillzyusko Please tag reviewers to continue review here as this seems to be follow-up!

**Comment by kirillzyusko** [comment]
@cubuspl42 @shubham1206agra tagging you to continue your review here

cc @roryabraham

**Comment by cubuspl42** [comment]
I remind about a few minor comments from the old PR

**Comment by kirillzyusko** [comment]
> I remind about a few minor comments from the old PR

Done 🙌 Would you mind to have a look again? 👀

**Comment by cubuspl42** [comment]
> [x] I ran the tests on **all platforms** & verified they passed on:
> 
>     * [x]  Android: Native
> 
>     * [x]  Android: mWeb Chrome
> 
>     * [x]  iOS: Native
> 
>     * [x]  iOS: mWeb Safari
> 
>     * [x]  MacOS: Chrome / Safari
> 
>     * [x]  MacOS: Desktop

Recently, we had [a similar situation](https://github.com/Expensify/App/pull/37891#issuecomment-1999770227) where the code changes were _supposed_ not to affect other platforms, but actually broke one of them.

Will this be different here? We don't know yet.

I know the checklist can be a pain, but let's _try_ to actually follow it when it _could_ be in the right.

**Comment by cubuspl42** [comment]
Minor thing: also please change "PROPOSAL: GH_LINK_ISSUE(COMMENT)" to "PROPOSAL: N/A" or something like that

**Comment by kirillzyusko** [comment]
> Recently, we had https://github.com/Expensify/App/pull/37891#issuecomment-1999770227 where the code changes were supposed not to affect other platforms, but actually broke one of them.
>
> Will this be different here? We don't know yet.
>
> I know the checklist can be a pain, but let's try to actually follow it when it could be in the right.

Yeah, it's very good point - let me record and attach videos from other platforms 👍

**Comment by kirillzyusko** [comment]
> Yeah, it's very good point - let me record and attach videos from other platforms 👍

@cubuspl42 done, I've uploaded all videos!

**Comment by cubuspl42** [comment]
@kirillzyusko Can you reproduce this?

https://github.com/Expensify/App/assets/2590174/435ba88d-dd3d-48a7-ab18-e94aa94a4ca8

**Comment by kirillzyusko** [comment]
> Can you reproduce this?

Let me try 👀

**Comment by kirillzyusko** [comment]
@cubuspl42 I think I fixed the problem (thank you again for reporting it 🙌 ). I patched a dependency - would you mind to pull latest changes, re-install `node_modules` (or simply run `postinstall` script) and re-test it? 👀

**Comment by cubuspl42** [comment]
Conflicts 🙁

**Comment by kirillzyusko** [comment]
@cubuspl42 resolved 🙂

**Comment by cubuspl42** [comment]
Sorry to be picky, but I found something else. I put actual effort to build `upstream/main` in a separate worktree and I tested on two virtual devices side by side (feature branch vs `main`) and as far as I can tell, this **is** a regression.

https://github.com/Expensify/App/assets/2590174/5dc2836a-7423-45cd-aee9-563496511d50

**Comment by cubuspl42** [comment]
As I understand it, the general goal is to ensure that we don't cover the element which has the user's attention with an input panel (like a keyboard or custom keyboard-alike component).

In this aspect, I find this a regression, as on `main` the edited message always ends up directly over the keyboard (cleanly visible), no matter where it was (relatively) positioned on the screen within the scroll view before the interaction.

On the feature branch, the keyboard can nearly cover the edited message. It can be worked around by manually scrolling after the fact, but I don't think it's enough.

**Comment by cubuspl42** [comment]
For reference, the behavior on `main`:

https://github.com/Expensify/App/assets/2590174/accff997-5297-411f-b39d-0192812db47f

Please note that, in this case, I also start to edit a message on the bottom of the screen (partially obscured by the bottom device edge), but it ends up being cleanly pushed up over the keyboard.

**Comment by roryabraham** [comment]
Will wait for a conflict-free branch before running an AdHoc build here

File: patches/react-native-keyboard-controller+1.10.4.patch [hunk 1]
```diff
@@ -0,0 +1,13 @@
+diff --git a/node_modules/react-native-keyboard-controller/src/components/KeyboardAvoidingView/index.tsx b/node_modules/react-native-keyboard-controller/src/components/KeyboardAvoidingView/index.tsx
+index 9e73caf..0d5f143 100644
+--- a/node_modules/react-native-keyboard-controller/src/components/KeyboardAvoidingView/index.tsx
++++ b/node_modules/react-native-keyboard-controller/src/components/KeyboardAvoidingView/index.tsx
+@@ -81,7 +81,7 @@ const KeyboardAvoidingView = forwardRef<View, React.PropsWithChildren<Props>>(
+     const onLayoutWorklet = useCallback((layout: LayoutRectangle) => {
+       "worklet";
+ 
+-      if (keyboard.isClosed.value) {
++      if (keyboard.isClosed.value || initialFrame.value === null) {
```
Comment (roryabraham) [inline patches/react-native-keyboard-controller+1.10.4.patch:10]: @kirillzyusko isn't this a library you maintain? can we add this in the upstream?

Comment (kirillzyusko) [reply patches/react-native-keyboard-controller+1.10.4.patch]: @roryabraham yes, I did quick patch-package just to verify that it actually fixes the problem. It fixes, so I pushed this changes to upstream and published a new version to `npm` and update `package.json` 👀

**Comment by kirillzyusko** [comment]
> Will wait for a conflict-free branch before running an AdHoc build here

@roryabraham I resolved merge conflicts - may I kindly ask you to review these changes?

@cubuspl42 I think I fixed the problem 🤞 May I kindly ask you to check again?

**Comment by kirillzyusko** [comment]
Looks like reassure fails:

image

because it fails in `main` as well:

image

**Comment by cubuspl42** [comment]
Sometimes this can be waited out.

**Comment by cubuspl42** [comment]
@kirillzyusko Would you solve conflicts and merge `main`? Hopefully, it also resolves the "Reassure Performance Tests".

**Comment by cubuspl42** [comment]
@shubham1206agra Do you remember that you are co-assigned? This is non-trivial change, so maybe don't wait for each other, this doesn't need to be sequential 🙂

**Comment by kirillzyusko** [comment]
> Would you solve conflicts and merge main? Hopefully, it also resolves the "Reassure Performance Tests".

@cubuspl42 Done 🙌

**Comment by kirillzyusko** [comment]
> Can you check https://github.com/kirillzyusko/react-native-keyboard-controller/releases/tag/1.11.5 and maybe bump the version?

@shubham1206agra done 🙌 

> Extra space between keyboard and banner

Can you provide a little bit more info? How did you open this screen (i. e. you tried to search for people to start a new conversation)? Would be awesome if you could add an expected result/reference how it looks now.

**Comment by kirillzyusko** [comment]
@shubham1206agra thank you for bumping it. I've seen bugs that you reported but didn't have a time to look into them yet (was busy with e2e tests).

Also I think it's wort to mention that starting from tomorrow I'll go to vacation, so realistically I can return to this task only after 2 weeks 😔

**Comment by hannojg** [comment]
Kiryl will be OOO until 19.04 - Expect to continue on this PR the week after returning!

**Comment by cubuspl42** [comment]
Some checks are failing, just saying 👆

**Comment by kirillzyusko** [comment]
> Some checks are failing, just saying 👇

Yeah, I know, thanks ❤️ 

We've temporarily postponed a work on this PR, because since the new architecture was enabled - the animation on chat screen (resizing the content along with the keyboard) became unsynchronized with the keyboard (and animations are junky in this PR as well).

So the best solution would be to fix animation in a current codebase and then proceed with this PR, but we'll see how it goes 👀

**Comment by kirillzyusko** [comment]
A little bit context: I've noticed that I started to use both `git merge` and `git rebase` so decided to simplify a little bit git history and be consistent, so decided to revert last `git merge` and perform a `git rebase` (and solve all conflicts and then force push a branch).

I did it (git didn't show up any conflicts) and I pushed this branch. GitHub immediately closed the PR and showed +0/-0 changes) because branch got corrupted. With help of @shubham1206agra I restored my changes prior to final rebase, but the problem now is that GH doesn't allow to re-open this PR (even if I push new commits and there is an actual diff):

image

I had to open a new PR: https://github.com/Expensify/App/pull/42143

Sorry for inconvenience 😔 If you have some ideas on how to restore this PR I'll be happy to hear 🙌

**Pull request closed without merging**